fix: improve error messages for missing database tables

ivanauth opened the pull request:
When users start SpiceDB without running migrations first, they get a confusing error like:

```
unable to list namespaces: ERROR: relation "namespace_config" does not exist (SQLSTATE 42P01)
```

This doesn't tell them what went wrong or how to fix it.

Now the error message is clear and actionable:

```
datastore error: the database schema has not been initialized; please run "spicedb datastore migrate": ERROR: relation "namespace_config" does not exist (SQLSTATE 42P01)
```

This applies to all SQL datastores (postgres, mysql, crdb) and covers read operations, write operations, and statistics queries.

Closes #2749

miparnisari commented on internal/datastore/common/errors.go:
```diff
@@ -176,3 +176,32 @@ type RevisionUnavailableError struct {
+		err,
+		codes.FailedPrecondition,
+		spiceerrors.ForReason(
+			v1.ErrorReason_ERROR_REASON_UNSPECIFIED,
```
nit: please create the right error code

miparnisari commented on internal/datastore/postgres/common/errors.go:
```diff
@@ -19,6 +19,7 @@ const (
 	pgReadOnlyTransaction       = "25006"
 	pgQueryCanceled             = "57014"
 	pgInvalidArgument           = "22023"
+	pgMissingTable              = "42P01"
```
there are two constants defined elsewhere: `postgresMissingTableErrorCode`

please remove them in favor of this one

miparnisari commented on internal/datastore/postgres/revisions.go:
```diff
@@ -312,6 +312,9 @@ func createNewTransaction(ctx context.Context, tx pgx.Tx, metadata map[string]an
 	cterr := tx.QueryRow(ctx, sql, args...).Scan(&newXID, &newSnapshot, &timestamp)
 	if cterr != nil {
+		if wrappedErr := common.WrapMissingTableError(cterr); wrappedErr != nil {
+			return newXID, newSnapshot, timestamp, wrappedErr
```
```suggestion
			cterr = ....
```

miparnisari commented on internal/datastore/postgres/reader.go:
```diff
@@ -252,6 +264,9 @@ func (r *pgReader) LookupNamespacesWithNames(ctx context.Context, nsNames []stri
 	})
 	if err != nil {
+		if wrappedErr := pgxcommon.WrapMissingTableError(err); wrappedErr != nil {
+			return nil, wrappedErr
```
```suggestion
			err = wrappedErr
```

otherwise you lose the `errUnableToListNamespaces` below

miparnisari reviewed (changes requested):
Used Postman and tried all the APIs against CRDB and Spanner. Notes:
- we're missing this change for Spanner. You can use `docker-compose -f docker-compose.spanner.yaml up --build` to test. E.g. 
    - call to WriteSchema I get `"unable to list caveats: spanner: code = \"NotFound\", desc = \"Table not found: caveat\", requestID = \"1.23ff9b5081a896f8.1.1.70.1\"",`.
    - call to DeleteRelationships I get `"object definition doc not found"`. This line has to be updated to account for when the error description says "table" https://github.com/authzed/spicedb/blob/f41f64166fee87eb65077120e50f8ea6ebdc16f6/internal/datastore/spanner/reader.go#L300
- in many cases, the errors returned to the user are prefixed with `error encoding zedtoken: unable to read unique ID` which makes no sense to me, and probably not to the end user 😆  It comes from https://github.com/authzed/spicedb/blob/01d72a1387ad154c0d268e161e5fdc37788eca2a/pkg/zedtoken/zedtoken.go#L71-L74. i'm very tempted to remove that `error encoding zedtoken` and to clarify what the "unique ID" is
- in spanner, if you didn't run migrations, calling Watch api gives a HTTP 400. But against CRDB, the watch service is completely disabled: `unknown service authzed.api.v1.WatchService`, which means that you need to re-boot the server after you ran the migrations. I wonder if the server is supposed to be hable to handle migrations running at the same time ?

Anyway. Could you update https://github.com/authzed/spicedb/blob/3370198fda1678280ff81477b1d3eca2d3cbdecd/internal/datastore/crdb/crdb.go#L576 to set a different error message in this case?

- when you call ImportBulk, you get this

```json
{
    "code": 2,
    "message": "statement description failed: ERROR: relation \"relation_tuple\" does not exist (SQLSTATE 42P01)",
    "details": []
}
```

, seems like we missed an error handling here: https://github.com/authzed/spicedb/blob/3370198fda1678280ff81477b1d3eca2d3cbdecd/internal/datastore/postgres/common/bulk.go#L79

miparnisari commented:
I also wonder if instead of littering all the datastore implementations with error handling, we could just prevent the server from coming up online when you have never ran `migrate`.

E.g. here

https://github.com/authzed/spicedb/blob/70cd75785eea27d6be697db13952a4745270b199/internal/services/health/health.go#L112-L115

if `dsReady.message` includes `datastore is not migrated: currently at revision \"\"`, then we could tell people to run `migrate`. we'd need to write a generic test ensuring that all datastores return empty revision when they haven't been migrated. 

I'm sure there are more changes required than this, perhaps an update to `Server.Complete` to include this check, but @josephschorr does this sound reasonable?

ivanauth commented:
Squashed to 2 commits and addressed all feedback:

* Replaced `WrapMissingTableError` with `IsMissingTableError` and `NewSchemaNotInitializedError` across datastores to preserve caller context
* Added case-insensitive “table not found” matching for Spanner
* Switched readiness gate to `resenje.org/singleflight` to fix lint
* Error code remains `UNSPECIFIED` until api#159 lands

Zedtoken prefix and Watch API behavior seem like separate concerns - happy to tackle in follow-up PRs.

Unit test failure (`TestCertWatcherStart` in `pkg/x509util`) is flaky and unrelated.
